mgr/dashboard: document Prometheus' security model

p-se opened the pull request:
Fixes: https://tracker.ceph.com/issues/45377

Signed-off-by: Patrick Seidensal <pseidensal@suse.com>


<!--
Thank you for opening a pull request!  Here are some tips on creating
a well formatted contribution.

Please give your pull request a title like "[component]: [short description]"

This is the format for commit messages:

"""
[component]: [short description]

[A longer multiline description]

Fixes: [ticket URL on tracker.ceph.com, create one if necessary]
Signed-off-by: [Your Name] <[your email]>
"""

The Signed-off-by line is important, and it is your certification that
your contributions satisfy the Developers Certificate or Origin.  For
more detail, see SubmittingPatches.rst.

The component is the short name of a major daemon or subsystem,
something like "mon", "osd", "mds", "rbd, "rgw", etc. For ceph-mgr modules,
give the component as "mgr/<module name>" rather than a path into pybind.

For more examples, simply use "git log" and look at some historical commits.

This was just a quick overview.  More information for contributors is available here:
https://raw.githubusercontent.com/ceph/ceph/master/SubmittingPatches.rst

-->
## Checklist
- [x] References tracker ticket
- [ ] Updates documentation if necessary
- [ ] Includes tests for new functionality or reproducer for bug

---

<details>
<summary>Show available Jenkins commands</summary>

- `jenkins retest this please`
- `jenkins test classic perf`
- `jenkins test crimson perf`
- `jenkins test signed`
- `jenkins test make check`
- `jenkins test make check arm64`
- `jenkins test submodules`
- `jenkins test dashboard`
- `jenkins test dashboard backend`
- `jenkins test docs`
- `jenkins render docs`
- `jenkins test ceph-volume all`
- `jenkins test ceph-volume tox`

</details>

LenzGr commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,22 @@ The available iSCSI gateways must be defined using the following commands::
+  all time series information contained in the database, plus a variety of
+  operational/debugging information.
+  
+  Though, Prometheus' HTTP API is, by default, limited to read-only
+  operations. Configurations cannot be changed using the API and secrets are
```
How about simplifying this part?
```suggestion
  However, Prometheus' HTTP API is limited to read-only operations by default.
  No configurations can be changed using the API and secrets are
```

LenzGr commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,22 @@ The available iSCSI gateways must be defined using the following commands::
+  
+  Though, Prometheus' HTTP API is, by default, limited to read-only
+  operations. Configurations cannot be changed using the API and secrets are
+  not exposed. Also, Prometheus has some built-in measures to mitigate the
```
```suggestion
  not exposed. Moreover, Prometheus has some built-in measures to mitigate the
```

LenzGr commented:
LGTM, I left a few small enhancement suggestions. Since this chapter is about Grafana dashboards, it's somewhat irritating to begin it with a warning about Prometheus without giving some context about its role in this scenario.

s0nea commented:
jenkins render docs

ceph-jenkins commented:
Doc render available at http://docs.ceph.com/ceph-prs/34902/

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -1,9 +1,10 @@
-The Ceph dashboard makes use of prometheus, grafana, and related tools
-to store and visualize detailed metrics on cluster utilization and
-performance.  Ceph users have three options:
+The Ceph dashboard makes use of `Prometheus <https://prometheus.io/>`_,
```
```suggestion
The Ceph dashboard uses `Prometheus <https://prometheus.io/>`_,

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+
+.. note::
+
+  Please note that Prometheus' security model presumes that untrusted users
```
```suggestion
  Prometheus' security model presumes that untrusted users

asettle replied:
Since it's in a called out "note" already when rendered, the beginning point is superfluous wording - but that's up to personal preference :) so feel free to ignore this one if you prefer it the other way

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+.. note::
+
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
```
```suggestion
  have access to the Prometheus HTTP endpoint and logs. Untrusted users have access to

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
+  all time series information contained in the database, plus a variety of
```
What's "all time series information"?

p-se replied:
"all time series information" is all the (meta)data Prometheus collects about machines and the Ceph cluster through Prometheus exporters, namely Prometheus manager module and Node exporter (in our current case).

I took the sentence from the Prometheus documentation and I'm not sure if it's necessary to go into that detail here, though I may be wrong.

asettle replied:
:thinking: hmm, so it is the name of the data. Although without the context it sort of makes it seem.. well... odd. What do you think of this:

"Untrusted users have access to all the (meta)data Prometheus collects contained in the database..." (you could then explain that the data is called "all time series information" but the direct answer works fine, provides more context for those who haven't had a lot to do with Prometheus. I do leave that up to your call, the audience here should be aware, but you never know.

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
+  all time series information contained in the database, plus a variety of
+  operational/debugging information.
```
```suggestion
  operational and debugging information

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+  operational/debugging information.
+
+  However, Prometheus' HTTP API is limited to read-only operations. No
+  configurations can be changed using the API and secrets are not exposed.
```
```suggestion
  However, Prometheus' HTTP API is limited to read-only operations. Configurations
  can *not* be changed using the API, and secrets can not be exposed.

p-se replied:
I've partly adapted the sentence to the one you provided, though I'd like to keep the "are not exposed" part, as particular configuration secrets are not exposed but I'm not so sure about manually added configuration options.

asettle replied:
Works for me!

asettle commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +15,27 @@ performance.  Ceph users have three options:
+  denial of service attacks.
+
+  Please see `Prometheus' Security
+  model<https://prometheus.io/docs/operating/security/>` for more detailed
```
```suggestion
  model <https://prometheus.io/docs/operating/security/>` for more detailed

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+`Grafana <https://grafana.com/>`_ needs to request data from `Prometheus
+<https://prometheus.io/>`_. The Grafana dashboards we provide contain queries
```
For what? For enabling the mebedding of Grafana dashboards? if so, include that :) see suggestion below

p-se replied:
No, not for enabling the embedding of Grafana dashboards but for the dashboards itself to be displayed. Every widget in the dashboard requires information to be displayed in a widget. The queries contained in those dashboards are used by Grafana to retrieve the data from Prometheus. 

The thing I try to clarify here is that the dashboards/widgets contain Prometheus specific queries and can therefore not be used by another data source (or application) other than Prometheus. My explanation may wasn't as good as I intended it to be.

What do think of 
```
`Grafana`_ needs to request data from `Prometheus <https://prometheus.io/>`_.
The Grafana dashboards we provide contain queries that are specific to
Prometheus. Although Grafana can have other data sources, the Grafana
dashboards therefore only support Prometheus as data source.
```

I just added `therefore` in the last sentence.

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
 
+`Grafana <https://grafana.com/>`_ needs to request data from `Prometheus
+<https://prometheus.io/>`_. The Grafana dashboards we provide contain queries
```
```suggestion
<https://prometheus.io/>`_ to enable the embedding of Grafana Dashboards. 
The Grafana dashboards provided contain queries

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
+
+.. note::
+
+  Please note that Prometheus' security model presumes that untrusted users
```
```suggestion
  Prometheus' security model presumes that untrusted users

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
+.. note::
+
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
```
```suggestion
  have access to the Prometheus HTTP endpoint and logs. Untrusted users have access to

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
+
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
+  all time series information contained in the database, plus a variety of
```
Same question as above regarding the meaning of "all time series information"?

p-se replied:
Same answer as above :)

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
+  Please note that Prometheus' security model presumes that untrusted users
+  have access to the Prometheus HTTP endpoint and logs. They have access to
+  all time series information contained in the database, plus a variety of
+  operational/debugging information.
```
```suggestion
  operational and debugging information.

asettle commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +378,27 @@ The available iSCSI gateways must be defined using the following commands::
+  operational/debugging information.
+
+  However, Prometheus' HTTP API is limited to read-only operations. No
+  configurations can be changed using the API and secrets are not exposed.
```
```suggestion
  However, Prometheus' HTTP API is limited to read-only operations. Configurations
  can *not* be changed using the API and secrets can not be exposed.

p-se commented:
@LenzGr 

> LGTM, I left a few small enhancement suggestions. Since this chapter is about Grafana dashboards, it's somewhat irritating to begin it with a warning about Prometheus without giving some context about its role in this scenario.

I've added a few sentences to give a short introduction how Prometheus is related to all this. Thanks for the hint!

@asettle 

I've updated the documentation according to your suggestions and our discussion. Thanks for your help!

tspmelo commented:
jenkins test make check

tspmelo commented:
jenkins test dashboard

s0nea commented on doc/cephadm/monitoring.rst:
```diff
@@ -1,9 +1,9 @@
-The Ceph dashboard makes use of prometheus, grafana, and related tools
-to store and visualize detailed metrics on cluster utilization and
-performance.  Ceph users have three options:
+The Ceph dashboard uses `Prometheus <https://prometheus.io/>`_, `Grafana
```
In the dashboard documentation we write the word 'dashboard' with a capital 'D' when mentioning together with 'Ceph' - 'Ceph Dashboard'. We may want to do the same here?

```suggestion
The Ceph Dashboard uses `Prometheus <https://prometheus.io/>`_, `Grafana
```

LenzGr replied:
Actually, I would then drop the leading "the", too:
```suggestion
Ceph Dashboard uses `Prometheus <https://prometheus.io/>`_, `Grafana
```

s0nea commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +377,31 @@ The available iSCSI gateways must be defined using the following commands::
+
+  Please see `Prometheus' Security model
+  <https://prometheus.io/docs/operating/security/>` for more detailed
+  information
```
```suggestion
  information.
```

LenzGr commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +377,31 @@ The available iSCSI gateways must be defined using the following commands::
 
+`Grafana`_ requires data from Prometheus. Although Grafana can use other data
+sources, the Grafana dashboards we provide contain queries that are specific to
+`Prometheus <https://prometheus.io/>`_. Our Grafana dashboards therefore
```
Minor nit: move the URL to the first mentioning of Prometheus in Line 380 instead.

LenzGr commented on doc/mgr/dashboard.rst:
```diff
@@ -378,6 +377,31 @@ The available iSCSI gateways must be defined using the following commands::
+`Grafana`_ requires data from Prometheus. Although Grafana can use other data
+sources, the Grafana dashboards we provide contain queries that are specific to
+`Prometheus <https://prometheus.io/>`_. Our Grafana dashboards therefore
+require Prometheus as data source. The Ceph :ref:`mgr-prometheus` also only
```
```suggestion
require Prometheus as the data source. The Ceph :ref:`mgr-prometheus` also only
```

sebastian-philipp commented:
@LenzGr please merge this as soon as you're ok with this PR.

votdev commented on doc/cephadm/monitoring.rst:
```diff
@@ -14,8 +14,27 @@ performance.  Ceph users have three options:
+
+  Please see `Prometheus' Security model
+  <https://prometheus.io/docs/operating/security/>` for more detailed
+  information
```
```suggestion
  information.
```

s0nea commented on doc/cephadm/monitoring.rst:
```diff
@@ -1,9 +1,9 @@
-performance.  Ceph users have three options:
+Ceph Dashboard uses `Prometheus <https://prometheus.io/>`_, `Grafana
+<https://grafana.com/>`_, and related tools to store and visualize detailed
+metrics on cluster utilization and performance.  Ceph users have three options:
```
```suggestion
metrics on cluster utilization and performance. Ceph users have three options:
```
